samza-sql/src/main/java/org/apache/samza/sql/data/SamzaSqlRelMsgMetadata.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.sql.data;
+
+import java.io.Serializable;
+import java.time.Instant;
+import org.codehaus.jackson.annotate.JsonProperty;
+
+
+/**
+ * Metadata of Samza Sql Rel Message. Contains metadata about the corresponding event or
+ * relational row of a table. Used as member of the {@link SamzaSqlRelMessage}.
+ */
+public class SamzaSqlRelMsgMetadata implements Serializable {
+  /**
+   * The timestamp of when the events actually happened
+   * set by and copied from the event source
+   * TODO: copy eventTime through from source to RelMessage
+   */
+  @JsonProperty("eventTime")
+  private String eventTime;
+
+  /**
+   * the timestamp of when Samza App received the event
+   * TODO: arrivalTime is set during conversion from IME to SamzaMessage
+   */
+  @JsonProperty("arrivalTime")
+  private String arrivalTime;
+
+  /**
+   * the timestamp when SamzaSQL query starts processing the event
+   * set by the SamzaSQL Scan operator
+   */
+  @JsonProperty("scanTime")
+  private String scanTime;
+
+  public SamzaSqlRelMsgMetadata(@JsonProperty("eventTime") String eventTime, @JsonProperty("arrivalTime") String arrivalTime,
+      @JsonProperty("scanTime") String scanTime) {
+    this.eventTime = eventTime;
+    this.arrivalTime =arrivalTime;
+    this.scanTime = scanTime;
+  }
+
+  @JsonProperty("eventTime")
+  public String getEventTime() { return eventTime;}
+
+  public void setEventTime(String eventTime) {
+    this.eventTime = eventTime;
+  }
+
+  public boolean hasEventTime() { return eventTime != null && !eventTime.isEmpty(); }
+
+  @JsonProperty("arrivalTime")
+  public String getarrivalTime() { return arrivalTime;}
+
+  public void setArrivalTime(String arrivalTime) {
+    this.arrivalTime = arrivalTime;
+  }
+
+  public boolean hasArrivalTime() { return arrivalTime != null && !arrivalTime.isEmpty(); }
```

Comment:
maybe I'm paranoid, but do we need these checks for arrival time and scan time ? I understand event time could be set or not. The reason for my paranoia is because this is the prime path where each and every event will have to go through these checks in every operator.

Reply:
I had the same worry, and for now we do need to check if arrivalTime is set or not, until we do set the arrivalTIme. But even after we do set arrivalTime, we will check/validate it is set (as in scanTime case now)

Reply:
discussed offline: added comments on scanTime and arrivalTime checks that they need to be removed once code is stable, and that eventTime check needs to stay there because it's availability depends on source of event